camera/first_person_control.go

```diff
@@ -0,0 +1,260 @@
+// Rotate rotates the camera in place by the specified angles.
+func (fpc *FirstPersonControl) Rotate(thetaDelta, phiDelta float32) {
+
+	// Rotate in place by rotation only in the X and Y axis
+	r := fpc.cam.Rotation()
+	fpc.cam.SetRotation(r.X+phiDelta, r.Y+thetaDelta, r.Z)
+}
+
+// Move moves the camera the specified amount relative from the camera position based on its view direction.
+func (fpc *FirstPersonControl) Move(deltaX, deltaY, deltaZ float32) {
+
+	// TODO: The Translate calls handle relaitive camera movement well, however it "flies" in the direction
+	// pointed at rather than "walks" along the XZ plane as desired for this purpose. A new function
+	// for "flying" could be useful but still need to figure out movement without flying.
+
+	// Translate camera by deltas
+	fpc.cam.TranslateX(deltaX)
+	fpc.cam.TranslateY(deltaY)
+	fpc.cam.TranslateZ(deltaZ)
+}
+
+// onMouse is called when an OnMouseDown/OnMouseUp event is received.
+func (fpc *FirstPersonControl) onMouse(evname string, ev interface{}) {
+
+	// If nothing enabled ignore event
+	if fpc.enabled == FPNone {
+		return
+	}
+
+	switch evname {
+	case window.OnMouseDown:
+		gui.Manager().SetCursorFocus(fpc)
+		mev := ev.(*window.MouseEvent)
+		switch mev.Button {
+		case window.MouseButtonLeft: // Rotate
+			if fpc.enabled&FPRot != 0 {
+				fpc.state = fpStateRotate
+				fpc.rotStart.Set(mev.Xpos, mev.Ypos)
+			}
+		case window.MouseButtonMiddle: // Move vertical only
+			if fpc.enabled&FPMoveY != 0 {
+				fpc.state = fpStateMoveVertical
+				fpc.moveStart = math32.Vector3{X: 0, Y: mev.Ypos, Z: 0}
+			}
+		case window.MouseButtonRight: // Move horizontal only
+			if fpc.enabled&FPMoveXZ != 0 {
+				fpc.state = fpStateMoveHorizontal
+				fpc.moveStart = math32.Vector3{X: mev.Xpos, Y: 0, Z: mev.Ypos}
+			}
+		}
+	case window.OnMouseUp:
+		gui.Manager().SetCursorFocus(nil)
+		fpc.state = fpStateNone
+	}
+}
+
+// onCursor is called when an OnCursor event is received.
+func (fpc *FirstPersonControl) onCursor(evname string, ev interface{}) {
+
+	// If nothing enabled ignore event
+	if fpc.enabled == FPNone || fpc.state == fpStateNone {
+		return
+	}
+
+	mev := ev.(*window.CursorEvent)
+	switch fpc.state {
+	case fpStateRotate:
+		c := -2 * math32.Pi * fpc.RotSpeed / fpc.winSize()
+		fpc.Rotate(c*(mev.Xpos-fpc.rotStart.X),
+			c*(mev.Ypos-fpc.rotStart.Y))
+		fpc.rotStart.Set(mev.Xpos, mev.Ypos)
+	case fpStateMoveVertical:
+		fpc.Move(0, fpc.MoveSpeed*(mev.Ypos-fpc.moveStart.Y), 0)
+		fpc.moveStart = math32.Vector3{X: 0, Y: mev.Ypos, Z: 0}
+	case fpStateMoveHorizontal:
+		fpc.Move(fpc.MoveSpeed*(mev.Xpos-fpc.moveStart.X), 0, fpc.MoveSpeed*(mev.Ypos-fpc.moveStart.Z))
+		fpc.moveStart = math32.Vector3{X: mev.Xpos, Y: 0, Z: mev.Ypos}
+	}
+}
+
+// onScroll is called when an OnScroll event is received.
+func (fpc *FirstPersonControl) onScroll(evname string, ev interface{}) {
+
+	if fpc.enabled&FPMoveY != 0 {
+		sev := ev.(*window.ScrollEvent)
+		fpc.Move(0, fpc.KeyMoveSpeed*sev.Yoffset, 0)
```

Comment:
I think the scroll behavior should be like zooming in (scroll up) or out (scroll down), so this line should update the z coordinate (like KeyW and KeyS).

Reply:
I can do that, only reason I didn't is because currently have the right click hold as moving/strafing along XZ and thought the mouse scroll should be different axis.

Reply:
I think middle click hold should move along the XY plane (similar to how it works in the orbit control).

If we want to get fancy in the future, holding a modifier like Shift could constrain the movement to either X axis or Y axis based on the movement the user has done so far (like what Blender does), but no need to do that now.

Reply:
@danaugrs in orbit control middle click puts mouse cursor in Zoom state, so that lines up with mouse scroll also being used for zoom.

However in orbit the right click mode is the pan state for the XY plane right? I was thinking for first person view perhaps it made more since for right click mode to move about XY (like WASD movement). let me know if you want it exactly like orbit for right click mode.